src/main/java/org/myrobotlab/service/Ros.java

```diff
@@ -1,38 +1,316 @@
+    return ((RosConfig) config).subscriptions;
+  }
+
+  public List<String> getTopics() throws InterruptedException {
+
+    RosServiceCallback ret = rosCallService("/rosapi/topics");
+    if (ret == null) {
+      return null;
+    }
+    RosMsg msg = (RosMsg) ret.msg;
+    if (msg != null) {
+      List topics = (List) ((Map) msg.values).get("topics");
+      log.info(ret.toString());
+      return topics;
+    } else {
+      return new ArrayList<>();
+    }
+  }
+
+  public RosMsg publishRosMsg(RosMsg msg) {
+    return msg;
+  }
+
+  // FIXME - its a blocking call
+  public RosServiceCallback rosCallService(String service) {
+    try {
+      RosConfig c = (RosConfig) config;
+      RosServiceCallback callback = new RosServiceCallback();
+      String id = UUID.randomUUID().toString();
+      callback.id = id;
+      callbacks.put(id, callback);
+
+      RosMsg msg = new RosMsg();
+      msg.id = id;
+      msg.op = "call_service";
+      msg.service = service;
+
+      sendJson(CodecUtils.toJson(msg));
+      synchronized (callback) {
+        callback.wait(c.serviceCallTimeoutMs);
+      }
+      return callback; // callbacks.get(id);
+
+    } catch (TimeExceededException ex) {
+      warn("timeout exceeded on ros service call");
+    } catch (Exception e) {
+      error(e);
+    }
+    return null;
+  }
+
+  public void sendJson(String json) {
+    try {
+      if (client == null) {
+        error("client not connected");
+        return;
+      }
+      client.send(json);
+    } catch (Exception e) {
+      error(e);
+    }
+  }
+  
+  public void rosSendMsg(RosMsg msg) {
+    String json = CodecUtils.toJson(msg);
+    sendJson(json);
+  }
+
+  public void rosPublish(String topic, String json) {
+    RosMsg msg = new RosMsg();
+    msg.op = "publish";
+    msg.topic = topic;
+    msg.msg = CodecUtils.fromJson(json);    
+    String msgJson = CodecUtils.toJson(msg);
+    sendJson(msgJson);
+  }
+
+
+  public void rosSubscribe(String topic) {
+    try {
+      RosConfig c = (RosConfig) config;
+      RosMsg msg = new RosMsg();
+      msg.op = "subscribe";
+      msg.topic = topic;
+      sendJson(CodecUtils.toJson(msg));
+      if (c.subscriptions == null) {
+        c.subscriptions = new ArrayList<>();
+      }
+      c.subscriptions.add(topic);
+    } catch (Exception e) {
+      error(e);
+    }
+  }
+
+  public void rosUnsubscribe(String topic) {
+    try {
+      RosConfig c = (RosConfig) config;
+
+      RosMsg msg = new RosMsg();
+      msg.op = "unsubscribe";
+      msg.topic = topic;
+      client.send(CodecUtils.toJson(msg));
+      if (c.subscriptions == null) {
+        c.subscriptions = new ArrayList<>();
+      }
+      c.subscriptions.remove(topic);
+    } catch (Exception e) {
+      error(e);
+    }
+  }
+
+  @Override
+  public void onRemoteMessage(String uuid, String msg) {
+
+    try {
+
+      RosMsg rosMsg = CodecUtils.fromJson(msg, RosMsg.class);
+      if (rosMsg.id != null && rosMsg.service != null) {

```

Comment:
No null check for rosMsg here. If we expect to always get a valid RosMsg, then we can simplify `getTopics()` a good amount

Reply:
Sorry, what I meant was checking if rosMsg itself was null is not being done, and it probably should be